mgr/dashboard: Add progress step to multisite automation wizard

aaSharma14 opened the pull request:
Fixes: https://tracker.ceph.com/issues/67829
Fixes: https://tracker.ceph.com/issues/67830

Signed-off-by: Aashish Sharma <aasharma@redhat.com>


Multi-site with multi-cluster

https://github.com/user-attachments/assets/31583be1-3c4d-4c79-991d-539b22f439f7


Multi-site without multi-cluster

https://github.com/user-attachments/assets/b22dee94-6d3d-49ca-83e6-3c9d5d7828c7









<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

aaSharma14 commented:
jenkins test api

cloudbehl commented:
Make the progress as a component and use it upgrade module and here as well. Take steps and metadata of it as data and use it in module. 

Lets be consistent with progress pattern in upgrade module. It has a loading icon for progress  and tick mark when the upgrade is done.

aaSharma14 commented:
> @aaSharma14 just to understand this better since I am not really an expert in the python stuffs, can you explain what are all you doing in the python area so I can review this properly? There are so many changes in the GetterThread class which I am not sure what for. And same in the rgw_client.py. And neither the commit nor the PR descriptions mention anything about it.

@nizamial09 , as we discussed before , there was an issue on updating the task progress after each step where we see 'can't start new thread' error which makes the summary service inactive. The error happens in `get_daemons_and_pools` method in `rbd_mirroring.py` which uses `ViewCache` class from `tools.py`. ViewCache class includes starting a new thread whenever it is called and it is being called by the summary endpoint every 5sec. I have modified the code to use a `ThreadPoolExecutor` which prevents creating too many threads and manages a pool of threads, ensuring that threads are reused.

aaSharma14 commented:
jenkins retest this please

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
> > @aaSharma14 just to understand this better since I am not really an expert in the python stuffs, can you explain what are all you doing in the python area so I can review this properly? There are so many changes in the GetterThread class which I am not sure what for. And same in the rgw_client.py. And neither the commit nor the PR descriptions mention anything about it.
> 
> @nizamial09 , as we discussed before , there was an issue on updating the task progress after each step where we see 'can't start new thread' error which makes the summary service inactive. The error happens in `get_daemons_and_pools` method in `rbd_mirroring.py` which uses `ViewCache` class from `tools.py`. ViewCache class includes starting a new thread whenever it is called and it is being called by the summary endpoint every 5sec. I have modified the code to use a `ThreadPoolExecutor` which prevents creating too many threads and manages a pool of threads, ensuring that threads are reused.

I have reverted this change for now. The error went away by increasing the kcli resources to 4GiB RAM and increased CPUs

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-wizard/rgw-multisite-wizard.component.html:
```diff
@@ -251,3 +207,123 @@
+
+<ng-template #progressCompleteTemplate>
+  <div *ngIf="isMultiClusterConfigured; else exportTokenTemplate">
+    <div class="text-center text-success">
+      Multi-site replication setup is complete.
+    </div>
```
i18n

aaSharma14 replied:
Updated, Thanks

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-wizard/rgw-multisite-wizard.component.ts:
```diff
@@ -29,7 +32,7 @@ export class RgwMultisiteWizardComponent implements OnInit {
   currentStepSub: Subscription;
   permissions: Permissions;
-  stepTitles = ['Create Realm & Zonegroup', 'Create Zone', 'Select Cluster'];
+  stepTitles = ['Create Realm & Zonegroup', 'Create Zone', 'Select Cluster', 'Review']; // If multi-cluster is configured
```
i18n

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-wizard/rgw-multisite-wizard.component.ts:
```diff
@@ -105,10 +112,17 @@ export class RgwMultisiteWizardComponent implements OnInit {
       this.isMultiClusterConfigured = this.clusterDetailsArray.length > 0;
       if (!this.isMultiClusterConfigured) {
-        this.stepTitles = ['Create Realm & Zonegroup', 'Create Zone', 'Export Multi-site token'];
+        this.stepTitles = ['Create Realm & Zonegroup', 'Create Zone', 'Review']; // If multi-cluster is not configured
```
lets have this as a constant/enum and reuse above and here

aaSharma14 replied:
Done. Thanks

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,54 @@
+      {{ message }}
+    </h5>
+  </div>
+</div>
```
nit: no new line

afreen23 reviewed (changes requested):
![image](https://github.com/user-attachments/assets/b38da3ad-100a-43b1-8147-cdfffcdecb14)
In the wizard we must adapt the font size to as per wizard font size/normal.
Also we can check the carbon guideleines/components here.

aaSharma14 commented:
> ![image](https://private-user-images.githubusercontent.com/25664409/361397311-b38da3ad-100a-43b1-8147-cdfffcdecb14.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3MzQ3MDksIm5iZiI6MTcyNDczNDQwOSwicGF0aCI6Ii8yNTY2NDQwOS8zNjEzOTczMTEtYjM4ZGEzYWQtMTAwYS00M2IxLTgxNDctY2RmZmZjZGVjYjE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA0NTMyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxODYyNjNkYmQ5NjJiYzI3YmUzNjFiYTg2Y2UxNTE0NzBjZWM4ODNiNDBiM2UzNjJkOGFhOWM3ZmYyN2YzNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Y1CkGn16H4vmvcdoZMdV1-_PIWdtYlxVWKgZKpR7alw) In the wizard we must adapt the font size to as per wizard font size/normal. Also we can check the carbon guideleines/components here.

Thanks @afreen23 , I have updated the front size (using `h5` now, before it was `h3`). Let me know if its looks better.
![Screenshot from 2024-08-27 10-20-13](https://github.com/user-attachments/assets/90fd80ec-ba44-42b3-acd6-da1657d54436)

aaSharma14 commented:
jenkins test api

afreen23 reviewed (changes requested):
Just a reminder , we need to have a tracker for this, the commit is missing.

Thanks

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/enum/multisite-wizard-steps.enum.ts:
```diff
@@ -0,0 +1,19 @@
+export enum StepTitles {
+  CreateRealmAndZonegroup = 'Create Realm & Zonegroup',
+  CreateZone = 'Create Zone',
+  SelectCluster = 'Select Cluster',
+  Review = 'Review'
+}
```
these StepTitles are used in wizard component where i18n is applied

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.spec.ts:
```diff
@@ -0,0 +1,22 @@
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+});
```
Am not sure if I'l be able to do it for this PR, since its already big enough than was intended to.

afreen23 replied:
Lets have an issue in our jira board for fixing it in 8.1, thanks

aaSharma14 replied:
Created one - https://tracker.ceph.com/issues/67852

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.ts:
```diff
@@ -0,0 +1,21 @@
+  @Input() loading: boolean;
+  @Input() inProgress: boolean;
+  @Input() executingTaskName: string;
+  @Input() currentStep: string; // For upgradeStatus.which
+  @Input() servicesComplete: string; // For upgradeStatus.services_complete
+  @Input() progressText: string; // For upgradeStatus.progress
+  @Input() message: string; // For upgradeStatus.message
```
I think these are some extra props that are required for upgrade scenario specifically so had to add them in this component. Regarding carbon am not sure if we are taking it up in this release and since I haven't worked on carbonization efforts yet, so not sure will be able to complete it in time.

afreen23 replied:
Lets not do uprade in this PR as it requires a lot of thinking and work. Thanks

aaSharma14 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1069,82 +1070,21 @@ class SyncFlowTypes(Enum):
+        self.progress_id = str(uuid.uuid4())
+        self.progress_title = ''
+        self.progress_done = 0
+        self.progress_total = 2  # Total number of major steps
```
We are doing all the progress related calculations from the backend itself, we are just receiving progress update in the UI

afreen23 replied:
Okay I see we are using progress module, nice!

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+      <div class="mt-2">
+        <ngb-progressbar type="info"
+                         [value]="progress"
+                         [striped]="true"></ngb-progressbar>
```
we used to have an animation here which is present when the action is in progress and now when paused.
```suggestion
                         [striped]="true"
                         [animation]="!isPaused"></ngb-progressbar>
```

aaSharma14 replied:
Done

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+    <div class="text-center w-75">
+      <ng-container *ngIf="servicesComplete && servicesComplete.length > 0"
+                    i18n>
+          Finished upgrading:
```
this is generic so shouldn't we use generic message here?

aaSharma14 replied:
Done

nizamial09 commented on src/pybind/mgr/mgr_module.py:
```diff
@@ -2512,7 +2512,7 @@ def send_rgwadmin_command(self, args: List[str],
                 stdout=subprocess.PIPE,
                 stderr=subprocess.PIPE,
-                timeout=10,
+                timeout=60,
```
shall we reset this before we forget?

aaSharma14 replied:
Removed

nizamial09 replied:
?

nizamial09 commented on src/pybind/mgr/rgw/module.py:
```diff
@@ -406,4 +406,4 @@ def import_realm_token(self,
         placement_spec = placement.get('placement') if placement else None
         self.rgw_zone_create(zone_name, realm_token, port, placement_spec, start_radosgw,
-                             zone_endpoints)
+                             zone_endpoints, secondary_zone_period_retry_limit=10)
```
is this needed?

aaSharma14 replied:
Yes, this is needed. This was a part of the PR that added retries to the period commit, so we need to adadpt our existing code to adapt this additional parameter as well

afreen23 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1069,82 +1070,21 @@ class SyncFlowTypes(Enum):
+    def update_progress(self, step_title):
+        self.progress_title = 'Multisite-Setup: ' + step_title
+        progress = (self.progress_done / self.progress_total)
+        mgr.remote('progress', 'update', self.progress_id,
```
With this we are making this feature dependent on progress module enablement.
If progress module is off, we wont be able to see this working.
We can add a check in UI for that alongwith the enablement option of this module.

Also, if turning off progres smodule is not a common thing then we can wait till we hit a BZ :)

aaSharma14 replied:
I don't think users really care about enabling/disabling the progress module. Since other components like upgrade/rbd_mirroring depend on progress module as well for updates and we do not add any explicit checks there about if progress module is enabled or not, so we can follow similar approach here as well

afreen23 replied:
Alrighty, thanks for info

afreen23 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1179,51 +1123,79 @@ def get_updated_endpoints(endpoints):
                 self.import_realm_token_to_cluster(cluster_fsid, realm_name,
                                                    realm_token_info, username)
+            else:
+                self.update_progress("Realm Export Token fetched successfully")
+                mgr.remote('progress', 'complete', self.progress_id)
```
Instead of setting these two seperately, we can just pass `progress title` and `progress action` in `update_progress` function

aaSharma14 replied:
Thanks, updated the code

afreen23 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1069,82 +1070,21 @@ class SyncFlowTypes(Enum):
+        self.progress_done = 0
+        self.progress_total = 2  # Total number of major steps
+
+    def update_progress(self, step_title):
```
Although you are passing step but this is better param name
```suggestion
    def update_progress(self, progress_title):
```

aaSharma14 replied:
Done, Thanks

afreen23 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1302,21 +1279,103 @@ def check_user_in_second_cluster(self, cluster_url, cluster_token, username):
                 logger.info("User content in the second cluster: %s", user_content)
-                for user in user_content:
-                    if user['user_id'] == username:
+                if isinstance(user_content, list):
+                    if username in user_content:
```
```suggestion
                if isinstance(user_content, list) and username in user_content:
```

aaSharma14 replied:
Done, Thanks

afreen23 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1880,19 +1938,7 @@ def get_user_list(self, zoneName: str):
             user_list = out
         except SubprocessError as error:
             raise DashboardException(error, http_status_code=500, component='rgw')
-
-        if len(user_list) > 0:
-            for user_name in user_list:
-                rgw_user_info_cmd = ['user', 'info', '--uid', user_name, '--rgw-zone', zoneName]
-                try:
-                    exit_code, out, _ = mgr.send_rgwadmin_command(rgw_user_info_cmd)
-                    if exit_code > 0:
-                        raise DashboardException('Unable to get user info',
-                                                 http_status_code=500, component='rgw')
-                    all_users_info.append(out)
-                except SubprocessError as error:
-                    raise DashboardException(error, http_status_code=500, component='rgw')
-        return all_users_info
```
why this deletion ?

aaSharma14 replied:
This method is not used anywhere else in the code. As this was lying unused and we needed a similar method now for our use where we just need the users list and not the user info along with it, so decided to reuse this method itself.

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-wizard/rgw-multisite-wizard.component.html:
```diff
@@ -251,3 +208,124 @@
+  <ng-container *ngTemplateOutlet="progressTemplate"></ng-container>
+</ng-template>
+
+<ng-template #progressCompleteTemplate>
+  <div *ngIf="isMultiClusterConfigured; else exportTokenTemplate">
+    <div class="text-center text-success"
+         i18n>
+      Multi-site replication setup is complete.
+    </div>
+  </div>
+</ng-template>
```
We should pass this message to progress component only
`<cd-progress helperText="Multi-site replication setup is complete"  status="finished"/>`

aaSharma14 replied:
This is required in this component itself as we are showing different templates on the same page based on different conditions

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-wizard/rgw-multisite-wizard.component.html:
```diff
@@ -251,3 +208,124 @@
+</ng-template>
+
+<ng-template #progressTemplate>
+  <cd-progress  [progress]="executingTask?.progress"
+                [executingTaskName]="executingTask?.name?.replace('progress/', '')">
+  </cd-progress>
```
We can manage these two values easily in our component and then pass the two to progress.
This will also help us to maintain and debug
```suggestion
  <cd-progress  [state]="state"
                            [label]="currentLabel">
  </cd-progress>
```

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+<div class="d-flex flex-column justify-content-center align-items-center bold">
+  <ng-container>
+    <h3 class="text-center"
+        [ngClass]="{ 'mt-3': inProgress && isPaused }">
+      <i *ngIf="inProgress && isPaused; else spinningIcon"
```
if we see having both spinner and progress now is creating redundancy.

aaSharma14 replied:
Should be a part of carbonization efforts once we take it up

afreen23 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+
+      <!-- Progress Bar -->
+      <div class="mt-2">
+        <ngb-progressbar type="info"
```
make progress bar cover full length as text , see Ibm carbon example , looks better.
![Uploading image.png…]()

aaSharma14 replied:
Should be a part of carbonization efforts once we take it up

afreen23 reviewed (changes requested):
ideally we should not be using spinner + progress both.
We should be using only progress bar and use **indeterminate state**  as [carbon](https://react.carbondesignsystem.com/?path=/docs/components-progressbar--overview) does:

[Screencast from 2024-08-28 21-12-53.webm](https://github.com/user-attachments/assets/0e592d43-8ad0-4ce3-9741-0d6f472221bc)

I would suggest to use this [progress](https://react.carbondesignsystem.com/?path=/docs/components-progressbar--overview) component  if this  is not required in squid.

Else, lets do these things:
1. Avoid spinner and replicate such indeterminate state. In case thats not possible then lets show spinner only when progress value is not passed or 0 and no error occurred.
2. Make component generic enough accepting following props:

```
<cd-progress label="Progress bar label" helperText="Optional helper text" value={75} status={active}/>
```
label = should be top and exaplain whats the **task** 
helperText = below and can be anything extra we want user to know - percentage, time, etc.
value = the visual progress out of 100
status = active/finished/error just 3 states are enough and for each we can have a change of colour. (active=default dashboard for general, finished= default dashboard for successful, error= default dashboard for errors)

Any further props if required should be generic as above.

aaSharma14 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1069,82 +1070,21 @@ class SyncFlowTypes(Enum):
+        self.progress_total = 2  # Total number of major steps
+
+    def update_progress(self, step_title):
+        self.progress_title = 'Multisite-Setup: ' + step_title
```
This is required to identify that progress is coming from which task in the UI. Since we have many tasks in summary endpoint, this helps us identify that this particular task is coming from the Multisite Setup

afreen23 replied:
But we are showing progress in a multisite UI or update in update UI, so this should is being well handled, no ?

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+                         [value]="progress"
+                         [striped]="true"></ngb-progressbar>
+      </div>
+      <p class="card-text text-muted">
+        <span class="float-end">{{ progress || 0 }} %</span>
+      </p>
+    </div>
```
Should be a part of carbonization efforts once we take it up

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+        <span class="text-success">
+            {{ servicesComplete }}
+        </span>
+      </ng-container>
```
Updated this

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+      </p>
+    </div>
+
+    <h4 class="text-center m-2"
+        *ngIf="progressText">
+      {{ progressText }}
+    </h4>
```
Should be a part of carbonization efforts once we take it up

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/progress/progress.component.html:
```diff
@@ -0,0 +1,60 @@
+      {{ progressText }}
+    </h4>
+
+    <h5 class="text-center mt-2"
+        *ngIf="message">
+      {{ message }}
+    </h5>
```
This is a part of upgrade progress where we show these extra messages in the progress scenario. Am not sure if I should be changing the existing implementation. This PR is just to add progress to the multisite wizard and not causing any regression to the existing upgrade functionality

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-progress/upgrade-progress.component.html:
```diff
@@ -1,48 +1,15 @@
+                 [progressText]="upgradeStatus.progress"
+                 [message]="upgradeStatus.message"
+                 [isPaused]="upgradeStatus.is_paused">
+    </cd-progress>
```
As I mentioned before, upgrade scenario requires these extra arguments which are not required by other scenarios. To not change any existing functionality of the upgrade progress thing, these extra props are being added

afreen23 commented:
> Thanks @afreen23 for the suggestions. Since these changes would increase the scope of this PR as it affects both the components (rgw/upgrade), so its better to open a tracker and do it separately. since the wizard itself is not carbonized we should probably deal with that one initially. Also, since this work was started prior to the carbonization efforts so I think a separate PR should be good to cover the carbonization for wizard/progress components.

I am okay with not introducing carbon changes, and I am okay not to do upgrade changes in this PR _(we may have a tracker for that)_ 
* What we must be doing is making the progress component generic as thats what we have introduced in this PR as a common component. See this comment with some [ideas](https://github.com/ceph/ceph/pull/58719#issuecomment-2316855314)
And subsequently reuse in other places in other PRs.
* And using spinner alongwith the progress does not seem right. At least not show it all the time and show only when some indeterminate state(_when progress value is not passed or 0 and no error occurred_).
* Since these are 3 BZs now so at least a tracker mentioning these 3 required here.
or separate 3 trackers for each. And kindly update that in commit

aaSharma14 commented:
jenkins test make check

afreen23 reviewed (approved):
LGTM, thanks!
Just need to test this.

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test api

nizamial09 commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -66,7 +66,7 @@ def create(self, username, password, ttl: Optional[int] = None):
         except KeyError:
             fsid = ''
-        if max_attempt == 0 or mgr.ACCESS_CTRL_DB.get_attempt(username) < max_attempt:
+        if max_attempt == 0 or mgr.ACCESS_CTRL_DB.get_attempt(username) < max_attempt:  # pylint: disable=R1702,line-too-long # noqa: E501
```
why is this?

aaSharma14 replied:
There were some linting errors in make check because of this - `too many nested blocks`

nizamial09 commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -95,19 +95,25 @@ def create(self, username, password, ttl: Optional[int] = None):
                     if fsid in multicluster_config['config']:
                         existing_entries = multicluster_config['config'][fsid]
-                        if not any((entry['user'] == username or entry['cluster_alias'] == 'local-cluster') for entry in existing_entries):  # noqa E501 #pylint: disable=line-too-long
+                        for entry in existing_entries:
```
better name than entry and existing entries?

aaSharma14 replied:
Updated

nizamial09 commented on src/pybind/mgr/dashboard/controllers/rgw.py:
```diff
@@ -118,20 +119,38 @@ def migrate(self, daemon_name=None, realm_name=None, zonegroup_name=None, zone_n
+        result = rgw_service_manager_instance.configure_rgw_credentials()
+        return result
+
+    @RESTController.Collection(method='GET', path='/available-ports')
+    @allow_empty_body
+    # pylint: disable=W0102,W0613
+    def get_available_ports(self):
+        rgw_service_manager_instance = RgwServiceManager()
+        result = rgw_service_manager_instance.find_available_port()
+        return result
```
why do we need to find the available port? isn't it the responsibility of cephadm to schedule the daemon to a port which it sees that is free?

aaSharma14 replied:
No, its not being taken care of by cephadm at present, if you run a new rgw service it takes `port 80` by default every time.

nizamial09 replied:
Probably we should start a conversation with cephadm to improve this behavior. Cephadm is in charge of deploying and managing a service and we can try to improve the behavior by making it smart enough to find an available port for a daemon that it is trying to deploy if the given port is unavailable. Its a worthy conversation to have IMO.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/upgrade/upgrade-progress/upgrade-progress.component.html:
you'd be doing the reviewer a great kindness if you actually separated them into different commits. Its also a responsible thing to do because if there are any regression with the upgrade component we can isolate that issue to one of those commit and revert them easily if that's the last and only option which is available. But right now if something like that happen, we are kind of doomed.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.html:
```diff
@@ -17,27 +17,25 @@
-                      [tableActions]="migrateTableAction">
-    </cd-table-actions>
-  </ng-template>
+  <cd-table-actions class="btn-group mb-4 me-2"
+                    [permission]="permission"
+                    [selection]="selection"
+                    [tableActions]="multisiteReplicationActions">
+  </cd-table-actions>
+  <cd-table-actions *ngIf="showMigrateAndReplicationActions"
+                    class="btn-group mb-4 me-2 secondary"
+                    [permission]="permission"
+                    [btnColor]="'light'"
+                    [selection]="selection"
+                    [tableActions]="migrateTableAction">
+  </cd-table-actions>
+  <cd-table-actions *ngIf="!showMigrateAndReplicationActions"
+                    class="btn-group mb-4 me-2"
+                    [permission]="permission"
+                    [selection]="selection"
+                    [tableActions]="createTableActions"
+                    [primaryDropDown]="true">
+  </cd-table-actions>
```
is it intentional to show multisite setup button and create realm button together?

aaSharma14 replied:
Yes, these buttons can co-exist, you can run through the wizard to create realm/zg/zone at once on single click. If a user wants to create a new realm/zg/zone individually that can be done as well.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/enum/multisite-wizard-steps.enum.ts:
```diff
@@ -0,0 +1,19 @@
+export enum StepTitles {
+  CreateRealmAndZonegroup = 'Create Realm & Zonegroup',
+  CreateZone = 'Create Zone',
+  SelectCluster = 'Select Cluster',
+  Review = 'Review'
+}
+
+export const STEP_TITLES_MULTI_CLUSTER_CONFIGURED = [
+  StepTitles.CreateRealmAndZonegroup,
+  StepTitles.CreateZone,
+  StepTitles.SelectCluster,
+  StepTitles.Review
+];
+
+export const STEP_TITLES_SINGLE_CLUSTER = [
+  StepTitles.CreateRealmAndZonegroup,
+  StepTitles.CreateZone,
+  StepTitles.Review
+];
```
why do we need enums and const for the step titles? this won't be reusable anywhere so i don't see the need for this to exist in a shared module. Maybe more fit in the rgw itself if you really think the enums and const are needed.

aaSharma14 replied:
Moved to multisite-wizard component

nizamial09 commented on src/pybind/mgr/dashboard/services/rgw_client.py:
```diff
@@ -1157,11 +1100,21 @@ def replace_hostname(self, endpoint, hostname_to_ip):
+        logger.info("cluster_details: %s", cluster_details)
+        if cluster_details:
+            cluster_details_dict = json.loads(cluster_details)
+            logger.info("cluster_details_dict: %s", cluster_details_dict)
```
we are registering a lot of critical sensitive details in the log which is a security risk when you log these stuffs. rather than dumping the entire dict of items in the log, just focus on the items you want or the user care more like the cluster name, the gateway name, the user name and not the entire dict where it contains things like cluster token in some cases and rgw tokens in other.

aaSharma14 replied:
Were just added for debugging purposes, forgot to remove, removed now

nizamial09 reviewed (commented):
Thank you @aaSharma14 this is a huge effort given that you already fixed multiple issues in different back-end (rgw module and cephadm) components to make this PR possible so thank you for that and your thorough testing.

My local testing was also good and worked as expected

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test make check

nizamial09 commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -94,20 +94,26 @@ def create(self, username, password, ttl: Optional[int] = None):
-                        existing_entries = multicluster_config['config'][fsid]
-                        if not any((entry['user'] == username or entry['cluster_alias'] == 'local-cluster') for entry in existing_entries):  # noqa E501 #pylint: disable=line-too-long
-                            existing_entries.append({
+                        cluster_configurations = multicluster_config['config'][fsid]
+                        for config_item in cluster_configurations:
+                            if config_item['user'] == username or config_item['cluster_alias'] == 'local-cluster':  # noqa E501  #pylint: disable=line-too-long
+                                config_item['token'] = token  # Update token
+                                break
```
checking this again, how's this related to the multi-site progress?

aaSharma14 replied:
we need to store the local cluster's token as well, lets suppose you have multi-cluster setup and you want to perform replication from cluster2 - local-cluster, in that case we switch to cluster2 from local and open the replication wizard, from there we can select local cluster from a list of clusters and do the replication

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test api
